packages/nest/src/utils/ast-utils.ts

```diff
@@ -0,0 +1,480 @@
+  );
+  // TODO: Check if we don't already import '@nestjs/graphql'
+  if (moduleImports.length !== 1) {
+    throw new Error(`main.ts can only import a single module`);
+  }
+  const moduleImport = moduleImports[0];
+  const moduleClassName = moduleImport.bindings.filter((b) =>
+    b.endsWith('Module')
+  )[0];
+
+  const modulePath = `${path.join(
+    path.dirname(mainPath),
+    moduleImport.moduleSpec
+  )}.ts`;
+  if (!host.exists(modulePath)) {
+    throw new Error(`Cannot find '${modulePath}'`);
+  }
+
+  const moduleSourceText = host.read(modulePath)!.toString('utf-8');
+  const moduleSource = ts.createSourceFile(
+    modulePath,
+    moduleSourceText,
+    ts.ScriptTarget.Latest,
+    true
+  );
+
+  return {
+    moduleSpec: moduleImport.moduleSpec,
+    mainPath,
+    modulePath,
+    moduleSource,
+    moduleClassName,
+    resolverClassName: moduleClassName.replace('Module', 'Resolver'),
+    resolverFileName: moduleClassName.replace('Module', '').toLowerCase(),
+  };
+}
+
+export interface BootstrapInfo {
+  moduleSpec: string;
+  modulePath: string;
+  mainPath: string;
+  moduleClassName: string;
+  moduleSource: ts.SourceFile;
+  resolverClassName: string;
+  resolverFileName: string;
+}
+
+export function getLibraryModule(host: Tree, config: any): BootstrapInfo {
+  const barrelExport = getLibraryBarrelExport(host, config);
+  let mainPath;
+  try {
+    mainPath = path.join(config.sourceRoot, barrelExport);
+  } catch (e) {
+    throw new Error('Module cannot be located ' + mainPath);
+  }
+
+  if (!host.exists(mainPath)) {
+    throw new Error('Module cannot be located');
+  }
+
+  const mainSource = host.read(mainPath)!.toString('utf-8');
+  const main = ts.createSourceFile(
+    mainPath,
+    mainSource,
+    ts.ScriptTarget.Latest,
+    true
+  );
+
+  const moduleExports = getExport(
+    main,
+    (s: string) => s.indexOf('Module') > -1
+  );
+
+  // // TODO: Check if we don't already import '@nestjs/graphql'
+  if (moduleExports.length > 1) {
+    throw new Error(`Class can only import a single module`);
+  }
+  const moduleExport = moduleExports[0];
+  const moduleClassName = moduleExport.bindings.filter((b) =>
+    b.endsWith('Module')
+  )[0];
+
+  return {
+    moduleSpec: barrelExport,
+    mainPath,
+    modulePath: mainPath,
+    moduleSource: main,
+    moduleClassName,
+    resolverClassName: moduleClassName.replace('Module', 'Resolver'),
+    resolverFileName: moduleClassName.replace('Module', '').toLowerCase(),
+  };
+}
+
+export function readBootstrapInfo(host: Tree, app: string): BootstrapInfo {
+  const config = getProjectConfig(host, app);
+
+  switch (config.projectType) {
+    case 'application':
+      return getApplicationModule(host, config);
+    case 'library':
+      return getLibraryModule(host, config);
+    default:
+      throw new Error(
+        `This schematic only works on an 'application' or 'library', not '${config.projectType}'`
+      );
+  }
+}
+
```

Comment:
The code below is take from the `@nrwl/angular` schematic and adjusted to work for Nest. It can probably be cleaned up or abstracted out.

Reply:
let's not copy the whole file. You can use a deep import to get reusable pieces since we're in the same repo :)